stablehlo/reference/ProcessGrid.cpp

```diff
@@ -0,0 +1,128 @@
+}
+
+SmallVector<std::pair<ProcessId, Tensor>> Process::rendezvous(
+    ProcessGroup processGroup, int64_t channelId, const Tensor &operand) {
+  return grid->rendezvous(processGroup, channelId, id, operand);
+}
+
+ProcessGrid::ProcessGrid(uint32_t numReplicas, uint32_t numPartitions)
+    : numReplicas_(numReplicas),
+      numPartitions_(numPartitions),
+      resourceLockMap_(
+          std::map<std::pair<ProcessGroup, int64_t>, std::mutex>()),
+      resourceConditionMap_(std::map<std::pair<ProcessGroup, int64_t>,
+                                     std::condition_variable>()),
+      channels_(std::map<std::pair<ProcessGroup, int64_t>,
+                         SmallVector<std::pair<ProcessId, Tensor>>>()) {}
+
+ProcessGroups ProcessGrid::crossReplica(
+    SmallVector<SmallVector<uint32_t>> replicaGroups) {
+  ProcessGroups processGroups;
+  for (auto replicaGroup : replicaGroups) {
+    for (uint32_t partitionId = 0; partitionId < numPartitions_;
+         ++partitionId) {
+      ProcessGroup processGroup;
+      for (uint32_t replicaId : replicaGroup)
+        processGroup.push_back({replicaId, partitionId});
+      processGroups.push_back(processGroup);
+    }
+  }
+  return processGroups;
+}
+
+ProcessGroups ProcessGrid::crossPartition(
+    SmallVector<SmallVector<uint32_t>> partitionGroups) {
+  ProcessGroups processGroups;
+  for (auto partitionGroup : partitionGroups) {
+    for (uint32_t replicaId = 0; replicaId < numReplicas_; ++replicaId) {
+      ProcessGroup processGroup;
+      for (uint32_t partitionId : partitionGroup)
+        processGroup.push_back({replicaId, partitionId});
+      processGroups.push_back(processGroup);
+    }
+  }
+  return processGroups;
+}
+
+SmallVector<std::pair<ProcessId, Tensor>> ProcessGrid::rendezvous(
+    ProcessGroup processGroup, int64_t channelId, ProcessId &processId,
+    const Tensor &operand) {
+  {
+    std::unique_lock<std::mutex> lk(
+        resourceLockMap_[{processGroup, channelId}]);
+    channels_[{processGroup, channelId}].push_back({processId, operand});
+  }
+
+  resourceConditionMap_[{processGroup, channelId}].notify_all();
+
+  {
+    std::unique_lock<std::mutex> lk(
+        resourceLockMap_[{processGroup, channelId}]);
+    if (channels_[{processGroup, channelId}].size() == processGroup.size())
+      resourceConditionMap_[{processGroup, channelId}].notify_all();
+
+    if (!resourceConditionMap_[{processGroup, channelId}].wait_for(
+            lk, std::chrono::seconds(3), [&] {
+              return channels_[{processGroup, channelId}].size() ==
+                     processGroup.size();
+            }))
+      llvm::report_fatal_error("rendezvous timed out");
+
+    std::sort(channels_[{processGroup, channelId}].begin(),
+              channels_[{processGroup, channelId}].end(),
+              [](std::pair<ProcessId, Tensor> &lhs,
+                 std::pair<ProcessId, Tensor> &rhs) {
+                return lhs.first < rhs.first;
+              });
```

Comment:
I don't think the sort piece is needed at the moment. And in the future it could be hidden as an implementation detail of `RendezvousResult`.